File: src/Java/601-700/637.TreeAverageOfLevels.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.Queue;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        LinkedList<Double> r = new LinkedList<>();
        if (root == null)
            return r;
        LinkedList<TreeNode> q = new LinkedList<>();
        q.add(root);
        while (q.size() > 0){
            double sum = 0;
            LinkedList<TreeNode> newq = new LinkedList<>();
            double l = q.size();
            for (TreeNode node : q){
                sum += node.val;
                if (node.left != null)
                    newq.add(node.left);
                if (node.right != null)
                    newq.add(node.right);
            }
            r.add(sum / l);
            q = newq;
        }
        return r;
    }
}

File: src/Python/601-700/637.TreeAverageOfLevels.py

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def averageOfLevels(self, root):
        """
        :type root: TreeNode
        :rtype: List[float]
        """
        r = []
        if root is None:
            return r
        q = [root]
        while len(q) > 0:
            sum = 0
            newq = []
            l = len(q)
            for node in q:
                sum += node.val
                if node.left is not None:
                    newq.append(node.left)
                if node.right is not None:
                    newq.append(node.right)
            r.append(sum / l)
            q = newq
        return r

if __name__ == '__main__':
    solution = Solution()
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    print(solution.averageOfLevels(root))
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(4)
    print(solution.averageOfLevels(root))
else:
    pass
